In [1]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library


### Scape wiki page to atrieve dataframe of boroughs and neighborhoods

In [2]:
#access wikipage and see page structure
from bs4 import BeautifulSoup
import requests
source=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(source, 'lxml')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className=document.documentElement.className.replace(/(^|\s)client-nojs(\s|$)/,"$1client-js$2");RLCONF={"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":900271985,"wgRevisionId":900271985,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":!1,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June",

In [3]:
#locate info in html <table class="wikitable sortable">
neighborhood = soup.find('table', class_ = 'wikitable')
neighborhood_rows = neighborhood.find_all('tr')
neighborhood_rows[:20]

[<tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighbourhood
 </th></tr>, <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>, <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>, <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
 </td></tr>, <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a>
 </td></tr>, <tr>
 <td>M6A</td>
 <td

In [4]:
#create a rough list to append rows
rough_list = []
for row in neighborhood_rows:
    info = row.text.split('\n')[1:-1] # remove empty str (the first and last items)
    rough_list.append(info)
rough_list[0:20] #preview the first 20 rows

[['Postcode', 'Borough', 'Neighbourhood'],
 ['M1A', 'Not assigned', 'Not assigned'],
 ['M2A', 'Not assigned', 'Not assigned'],
 ['M3A', 'North York', 'Parkwoods'],
 ['M4A', 'North York', 'Victoria Village'],
 ['M5A', 'Downtown Toronto', 'Harbourfront'],
 ['M5A', 'Downtown Toronto', 'Regent Park'],
 ['M6A', 'North York', 'Lawrence Heights'],
 ['M6A', 'North York', 'Lawrence Manor'],
 ['M7A', "Queen's Park", 'Not assigned'],
 ['M8A', 'Not assigned', 'Not assigned'],
 ['M9A', 'Etobicoke', 'Islington Avenue'],
 ['M1B', 'Scarborough', 'Rouge'],
 ['M1B', 'Scarborough', 'Malvern'],
 ['M2B', 'Not assigned', 'Not assigned'],
 ['M3B', 'North York', 'Don Mills North'],
 ['M4B', 'East York', 'Woodbine Gardens'],
 ['M4B', 'East York', 'Parkview Hill'],
 ['M5B', 'Downtown Toronto', 'Ryerson'],
 ['M5B', 'Downtown Toronto', 'Garden District']]

In [5]:
#turn the rough_list to pandas Dataframe
neighbor_df=pd.DataFrame(rough_list[1:], columns=rough_list[0])
# Re-spell Postcode as PostCode and Neighbourhood as Neighborhood
neighbor_df = neighbor_df.rename(columns={neighbor_df.columns[0]: "PostCode" ,
                                         neighbor_df.columns[2]: "Neighborhood"})
neighbor_df.head()

PostCode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [6]:
#drop any rows with "Not assigned" in Borough
neighbor_df = neighbor_df[neighbor_df.Borough != 'Not assigned']
neighbor_df.reset_index(drop=True, inplace=True)
neighbor_df.head()

PostCode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights

In [7]:
#create a new df group by PostCode
grouped = neighbor_df.groupby(['PostCode'])
#combine Neighborhood with same PostCode in a new df
neighbor_grouped=grouped['Neighborhood'].sum()
# adds spaces and commas between neighborhoods
neighbor_grouped = grouped['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
neighbor_grouped.head()

PostCode
M1B                            Rouge, Malvern
M1C    Highland Creek, Rouge Hill, Port Union
M1E         Guildwood, Morningside, West Hill
M1G                                    Woburn
M1H                                 Cedarbrae
Name: Neighborhood, dtype: object

In [8]:
#'Not Assigned' Neighborhodd updated as Borough
# matches a borough to each postcode
borough_grouped = grouped['Borough'].apply(lambda x: set(x).pop())
# turn borough_grouped and neighbor_grouped into dataframes
borough = borough_grouped.to_frame()
neighbor = neighbor_grouped.to_frame()
#combine the dataframe borough and the dataframe neighborhood into one dataframe
combined_df = borough.merge(neighbor, on="PostCode")
combined_df.reset_index(drop=False, inplace=True)
combined_df.head()

PostCode      Borough                            Neighborhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [9]:
#No. of rows and columns in combined_df
print("{} rows and {} columns in Toronto neighborhoods".format((combined_df.shape)[0], (combined_df.shape)[1]))

103 rows and 3 columns in Toronto neighborhoods


### Create full table with latitude and longitude

In [10]:
#read geo coordinates csv files instead of using Geospatial_data
geo_df=pd.read_csv("Geospatial_Coordinates.csv")
#rename "Postal Code" in geo_df as "PostCode" to match combined_df
geo_df=geo_df.rename(columns={geo_df.columns[0]: "PostCode"})
#combine geo_df and combined_df 
ready_df=combined_df.merge(geo_df, on="PostCode")
ready_df

PostCode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.23

### Explore and cluster the neighborhoods in Toronto

In [11]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(ready_df['Borough'].unique()),
        ready_df.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


In [12]:
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

WARNING conda.base.context:use_only_tar_bz2(632): Conda is constrained to only using the old .tar.bz2 file format because you have conda-build installed, and it is <3.18.3.  Update or remove conda-build to get smaller downloads and faster extractions.
Solving environment: done

# All requested packages already installed.

WARNING conda.base.context:use_only_tar_bz2(632): Conda is constrained to only using the old .tar.bz2 file format because you have conda-build installed, and it is <3.18.3.  Update or remove conda-build to get smaller downloads and faster extractions.
Solving environment: done

# All requested packages already installed.

Folium installed
Libraries imported.


In [13]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [14]:
from geopy.exc import GeocoderTimedOut

def do_geocode(address):
    try:
        return geopy.geocode(address)
    except GeocoderTimedOut:
        return do_geocode(address)

##### Use geopy library to get the latitude and longitude values of Toronto

In [15]:
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address, timeout=10)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


##### Create a map of Toronto with neighborhoods superimposed on top.

In [16]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(ready_df['Latitude'], ready_df['Longitude'], ready_df['Borough'], ready_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

##### Above map using blue pop-ups to show 103 neighborhoods in Toronto. More are concentrated in the center of city

##### Cluster by borough and visualize

In [17]:
borough_list = ready_df['Borough'].unique()
borough_list

array(['Scarborough', 'North York', 'East York', 'East Toronto',
       'Central Toronto', 'Downtown Toronto', 'York', 'West Toronto',
       "Queen's Park", 'Mississauga', 'Etobicoke'], dtype=object)

In [19]:
#find lat and log for all boroughs
def borough_loc(list_of_places):
    for place in list_of_places:
        address = (place + ", Ontario, Canada")
        geolocator = Nominatim(user_agent="TO_explorer", timeout=10)
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
        print('{''}, {}, {},'.format(place,latitude,longitude))

borough_loc(borough_list)

Scarborough, 43.773077, -79.257774,
North York, 43.7708175, -79.4132998,
East York, 43.6913391, -79.3278212,
East Toronto, 43.6247901, -79.3934918,
Central Toronto, 43.653963, -79.387207,
Downtown Toronto, 43.655115, -79.380219,
York, 44.0007518, -79.4372217,
West Toronto, 43.653963, -79.387207,
Queen's Park, 43.6599803, -79.3903686,
Mississauga, 43.590338, -79.645729,
Etobicoke, 43.67145915, -79.5524920661167,


In [20]:
#create Boroughs dataframe
borough_place = ['Scarborough', 43.773077, -79.257774,
'North York', 43.7708175, -79.4132998,
'East York', 43.6913391, -79.3278212,
'East Toronto', 43.6247901, -79.3934918,
'Central Toronto', 43.653963, -79.387207,
'Downtown Toronto', 43.655115, -79.380219,
'York', 44.0007518, -79.4372217,
'West Toronto', 43.653963, -79.387207,
"Queen's Park", 43.6599803, -79.3903686,
'Mississauga', 43.590338, -79.645729,
'Etobicoke', 43.67145915, -79.5524920661167,]
borough_df = pd.DataFrame(np.array(borough_place).reshape(11, 3), columns=['Borough', 'Latitude', 'Longitude'])
borough_df

Borough     Latitude          Longitude
0        Scarborough    43.773077         -79.257774
1         North York   43.7708175        -79.4132998
2          East York   43.6913391        -79.3278212
3       East Toronto   43.6247901        -79.3934918
4    Central Toronto    43.653963         -79.387207
5   Downtown Toronto    43.655115         -79.380219
6               York   44.0007518        -79.4372217
7       West Toronto    43.653963         -79.387207
8       Queen's Park   43.6599803        -79.3903686
9        Mississauga    43.590338         -79.645729
10         Etobicoke  43.67145915  -79.5524920661167

In [21]:
#change data type of latitude and longitude from object to float
borough_df.dtypes
borough_df['Latitude']=borough_df['Latitude'].astype('float')
borough_df['Longitude']=borough_df['Longitude'].astype('float')

In [22]:
# create a map of Toronto
map_toronto_boroughs = folium.Map(location = [43.653963, -79.387207], zoom_start = 10)

#add neighborhood markers to the Toronto map
for lat, long, bor in zip(borough_df['Latitude'], borough_df['Longitude'], 
                                 borough_df['Borough']):
    label = '{}'.format(bor)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius = 7, 
        popup = label,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto_boroughs)
        
map_toronto_boroughs

##### find out the most populated neighborhood

In [23]:
ready_df['Borough'].value_counts()

North York          24
Downtown Toronto    18
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East Toronto         5
East York            5
Queen's Park         1
Mississauga          1
Name: Borough, dtype: int64

###### It seems that North York has most neighborhoods. and I want to explore more on this part.

In [24]:
#create a dataframe for North York
northyork_data = ready_df[ready_df['Borough'] == 'North York'].reset_index(drop=True)
northyork_data.head()

PostCode     Borough                  Neighborhood   Latitude  Longitude
0      M2H  North York             Hillcrest Village  43.803762 -79.363452
1      M2J  North York  Fairview, Henry Farm, Oriole  43.778517 -79.346556
2      M2K  North York               Bayview Village  43.786947 -79.385975
3      M2L  North York      Silver Hills, York Mills  43.757490 -79.374714
4      M2M  North York       Newtonbrook, Willowdale  43.789053 -79.408493

In [25]:
northyork_data.shape

(24, 5)

In [26]:
address = 'North York, TO'

geolocator = Nominatim(user_agent="TO_explorer",timeout=10)
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of North York are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of North York are 43.7708175, -79.4132998.


In [27]:
# create map of North York using latitude and longitude values
map_northyork = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(northyork_data['Latitude'], northyork_data['Longitude'], northyork_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_northyork)  
    
map_northyork

##### Play with Foursquare

In [28]:
# Define Foursquare Credentials and Version
CLIENT_ID = 'QKQFEDFKSRYUUADF4II0Z45RSVCZS5UPSBYIW3BP34M2TRRG' # your Foursquare ID
CLIENT_SECRET = '1WWJEYCJCMBUF0PYCZ441JYX2LHDNRVHN32CCLGJANMJQRYJ' # your Foursquare Secret
VERSION = '20190704' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QKQFEDFKSRYUUADF4II0Z45RSVCZS5UPSBYIW3BP34M2TRRG
CLIENT_SECRET:1WWJEYCJCMBUF0PYCZ441JYX2LHDNRVHN32CCLGJANMJQRYJ


In [29]:
# Let's explore the first neighborhood in our dataframe
neighborhood_latitude = northyork_data.loc[1, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = northyork_data.loc[1, 'Longitude'] # neighborhood longitude value

neighborhood_name = northyork_data.loc[1, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Fairview, Henry Farm, Oriole are 43.7785175, -79.3465557.


In [30]:
# let's get the top 100 venues that are in Hillcrest Village within a radius of 500 meters.
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION, neighborhood_latitude, neighborhood_longitude, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=QKQFEDFKSRYUUADF4II0Z45RSVCZS5UPSBYIW3BP34M2TRRG&client_secret=1WWJEYCJCMBUF0PYCZ441JYX2LHDNRVHN32CCLGJANMJQRYJ&v=20190704&ll=43.7785175,-79.3465557&radius=500&limit=100'

In [31]:
#Send the GET request and examine the resutls
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d1eb27ee706550025026581'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Toronto',
  'headerFullLocation': 'Toronto',
  'headerLocationGranularity': 'city',
  'totalResults': 66,
  'suggestedBounds': {'ne': {'lat': 43.7830175045, 'lng': -79.34033481358881},
   'sw': {'lat': 43.7740174955, 'lng': -79.35277658641118}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e848fbb5c5c9240de8e6a80',
       'name': 'The LEGO Store',
       'location': {'address': '1800 Sheppard Ave E',
        'crossStreet': 'at Don Mills Rd',
        'lat': 43.77820727238842,
        'lng': -79.34348299621146,
        'labeledLatLngs': [{'label': 'display',
          

In [32]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [33]:
#clean the json and structure it into a pandas dataframe
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name        categories        lat        lng
0                    The LEGO Store  Toy / Game Store  43.778207 -79.343483
1  SilverCity Fairview Mall Cinemas     Movie Theater  43.778681 -79.344085
2                  CF Fairview Mall     Shopping Mall  43.777750 -79.344105
3                 Michel's Baguette            Bakery  43.777082 -79.344557
4            Hero Certified Burgers      Burger Joint  43.777295 -79.344584

In [34]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

66 venues were returned by Foursquare.


#### Explore Neighborhoods in North York

In [35]:
#create a function to repeat the same process to all the neighborhoods in North York
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [36]:
#write the code to run the above function on each neighborhood and create a new dataframe called "northyork_venues"
northyork_venues=getNearbyVenues(names=northyork_data['Neighborhood'], 
                                 latitudes=northyork_data['Latitude'], 
                                 longitudes=northyork_data['Longitude'] )


Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Bedford Park, Lawrence Manor East
Lawrence Heights, Lawrence Manor
Glencairn
Downsview, North Park, Upwood Park
Humber Summit
Emery, Humberlea


In [37]:
#check the size of the resulting dataframe
print(northyork_venues.shape)
northyork_venues.head()

(251, 7)


Neighborhood  Neighborhood Latitude  \
0             Hillcrest Village              43.803762   
1             Hillcrest Village              43.803762   
2             Hillcrest Village              43.803762   
3             Hillcrest Village              43.803762   
4  Fairview, Henry Farm, Oriole              43.778517   

   Neighborhood Longitude                   Venue  Venue Latitude  \
0              -79.363452  Eagle's Nest Golf Club       43.805455   
1              -79.363452         AY Jackson Pool       43.804515   
2              -79.363452            Villa Madina       43.801685   
3              -79.363452       Duncan Creek Park       43.805539   
4              -79.346556          The LEGO Store       43.778207   

   Venue Longitude            Venue Category  
0       -79.364186               Golf Course  
1       -79.366138                      Pool  
2       -79.363938  Mediterranean Restaurant  
3       -79.360695                   Dog Run  
4       -79.343483          Toy / Game Store

In [38]:
northyork_venues['Neighborhood'].value_counts()

Fairview, Henry Farm, Oriole                       66
Willowdale South                                   37
Bedford Park, Lawrence Manor East                  25
Flemingdon Park, Don Mills South                   23
Bathurst Manor, Downsview North, Wilson Heights    18
Lawrence Heights, Lawrence Manor                   16
Northwood Park, York University                     7
Downsview Northwest                                 5
York Mills West                                     5
Don Mills North                                     5
Willowdale West                                     5
Downsview, North Park, Upwood Park                  5
Glencairn                                           4
Hillcrest Village                                   4
Bayview Village                                     4
Victoria Village                                    4
Downsview Central                                   4
Parkwoods                                           3
Humber Summit               

In [39]:
#see how many venues for each neighborhood
northyork_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Bathurst Manor, Downsview North, Wilson Heights                     18   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   25   
CFB Toronto, Downsview East                                          3   
Don Mills North                                                      5   
Downsview Central                                                    4   
Downsview Northwest                                                  5   
Downsview West                                                       3   
Downsview, North Park, Upwood Park                                   5   
Emery, Humberlea                                                     1   
Fairview, Henry Farm, Oriole                                        66   
Flemingdon Park, Don Mills South                                    23   
Glencairn                                                            4   
Hillcrest Village                                                    4   
Humber Summit                                                        3   
Lawrence Heights, Lawrence Manor                                    16   
Northwood Park, York University                                      7   
Parkwoods                                                            3   
Silver Hills, York Mills                                             1   
Victoria Village                                                     4   
Willowdale South                                                    37   
Willowdale West                                                      5   
York Mills West                                                      5   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Bathurst Manor, Downsview North, Wilson Heights                      18   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    25   
CFB Toronto, Downsview East                                           3   
Don Mills North                                                       5   
Downsview Central                                                     4   
Downsview Northwest                                                   5   
Downsview West                                                        3   
Downsview, North Park, Upwood Park                                    5   
Emery, Humberlea                                                      1   
Fairview, Henry Farm, Oriole                                         66   
Flemingdon Park, Don Mills South                                     23   
Glencairn                                                             4   
Hillcrest Village                                                     4   
Humber Summit                                                         3   
Lawrence Heights, Lawrence Manor                                     16   
Northwood Park, York University                                       7   
Parkwoods                                                             3   
Silver Hills, York Mills                                              1   
Victoria Village                                                      4   
Willowdale South                                                     37   
Willowdale West                                                       5   
York Mills West                                                       5   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Bathurst Manor, Downsview North, Wilson Heights     18              18   
Bayview Village                                      4               4   
Bedford Park, Lawrence Man

In [40]:
#find out how many unique categories can be curated from all the returned venues
print('There are {} uniques categories.'.format(len(northyork_venues['Venue Category'].unique())))

There are 111 uniques categories.


In [41]:
northyork_venues.head()

Neighborhood  Neighborhood Latitude  \
0             Hillcrest Village              43.803762   
1             Hillcrest Village              43.803762   
2             Hillcrest Village              43.803762   
3             Hillcrest Village              43.803762   
4  Fairview, Henry Farm, Oriole              43.778517   

   Neighborhood Longitude                   Venue  Venue Latitude  \
0              -79.363452  Eagle's Nest Golf Club       43.805455   
1              -79.363452         AY Jackson Pool       43.804515   
2              -79.363452            Villa Madina       43.801685   
3              -79.363452       Duncan Creek Park       43.805539   
4              -79.346556          The LEGO Store       43.778207   

   Venue Longitude            Venue Category  
0       -79.364186               Golf Course  
1       -79.366138                      Pool  
2       -79.363938  Mediterranean Restaurant  
3       -79.360695                   Dog Run  
4       -79.343483          Toy / Game Store

#### Analyze Each Neighborhood

In [42]:
# one hot encoding
northyork_onehot = pd.get_dummies(northyork_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
northyork_onehot['Neighborhood'] = northyork_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [northyork_onehot.columns[-1]] + list(northyork_onehot.columns[:-1])
northyork_onehot = northyork_onehot[fixed_columns]

northyork_onehot.head()

Neighborhood  Accessories Store  Airport  \
0             Hillcrest Village                  0        0   
1             Hillcrest Village                  0        0   
2             Hillcrest Village                  0        0   
3             Hillcrest Village                  0        0   
4  Fairview, Henry Farm, Oriole                  0        0   

   American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0                    0                    0                 0   
1                    0                    0                 0   
2                    0                    0                 0   
3                    0                    0                 0   
4                    0                    0                 0   

   Athletics & Sports  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0                   0       0     0    0               0                 0   
1                   0       0     0    0               0                 0   
2                   0       0     0    0               0                 0   
3                   0       0     0    0               0                 0   
4                   0       0     0    0               0                 0   

   Beer Store  Bike Shop  Boutique  Bridal Shop  Bubble Tea Shop  \
0           0          0         0            0                0   
1           0          0         0            0                0   
2           0          0         0            0                0   
3           0          0         0            0                0   
4           0          0         0            0                0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Butcher  Cafeteria  \
0             0              0         0            0        0          0   
1             0              0         0            0        0          0   
2             0              0         0            0        0          0   
3             0              0         0            0        0          0   
4             0              0         0            0        0          0   

   Café  Candy Store  Caribbean Restaurant  Chinese Restaurant  \
0     0            0                     0                   0   
1     0            0                     0                   0   
2     0            0                     0                   0   
3     0            0                     0                   0   
4     0            0                     0                   0   

   Clothing Store  Coffee Shop  Comfort Food Restaurant  \
0               0            0                        0   
1               0            0                        0   
2               0            0                        0   
3               0            0                        0   
4               0            0                        0   

   Construction & Landscaping  Convenience Store  Cosmetics Shop  \
0                           0                  0               0   
1                           0                  0               0   
2                           0                  0               0   
3                           0                  0               0   
4                           0                  0               0   

   Cupcake Shop  Deli / Bodega  Department Store  Dessert Shop  \
0             0              0                 0             0   
1             0              0                 0             0   
2             0              0                 0             0   
3             0              0                 0             0   
4             0              0                 0             0   

   Dim Sum Restaurant  Diner  Discount Store  Dog Run  Electronics Store  \
0                   0      0               0        0                  0   
1                   0      0               0        0                  0   
2                   0      0               0        0                  0   
3                   0      0               0  

In [43]:
northyork_onehot.shape

(251, 112)

In [44]:
#group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
northyork_grouped = northyork_onehot.groupby('Neighborhood').mean().reset_index()
northyork_grouped

Neighborhood  Accessories Store  \
0   Bathurst Manor, Downsview North, Wilson Heights             0.0000   
1                                   Bayview Village             0.0000   
2                 Bedford Park, Lawrence Manor East             0.0000   
3                       CFB Toronto, Downsview East             0.0000   
4                                   Don Mills North             0.0000   
5                                 Downsview Central             0.0000   
6                               Downsview Northwest             0.0000   
7                                    Downsview West             0.0000   
8                Downsview, North Park, Upwood Park             0.0000   
9                                  Emery, Humberlea             0.0000   
10                     Fairview, Henry Farm, Oriole             0.0000   
11                 Flemingdon Park, Don Mills South             0.0000   
12                                        Glencairn             0.0000   
13                                Hillcrest Village             0.0000   
14                                    Humber Summit             0.0000   
15                 Lawrence Heights, Lawrence Manor             0.0625   
16                  Northwood Park, York University             0.0000   
17                                        Parkwoods             0.0000   
18                         Silver Hills, York Mills             0.0000   
19                                 Victoria Village             0.0000   
20                                 Willowdale South             0.0000   
21                                  Willowdale West             0.0000   
22                                  York Mills West             0.0000   

     Airport  American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0   0.000000             0.000000             0.000000          0.000000   
1   0.000000             0.000000             0.000000          0.000000   
2   0.000000             0.040000             0.000000          0.000000   
3   0.333333             0.000000             0.000000          0.000000   
4   0.000000             0.000000             0.000000          0.000000   
5   0.000000             0.000000             0.000000          0.000000   
6   0.000000             0.000000             0.000000          0.000000   
7   0.000000             0.000000             0.000000          0.000000   
8   0.000000             0.000000             0.000000          0.000000   
9   0.000000             0.000000             0.000000          0.000000   
10  0.000000             0.015152             0.000000          0.015152   
11  0.000000             0.000000             0.000000          0.086957   
12  0.000000             0.000000             0.000000          0.000000   
13  0.000000             0.000000             0.000000          0.000000   
14  0.000000             0.000000             0.000000          0.000000   
15  0.000000             0.000000             0.000000          0.000000   
16  0.000000             0.000000             0.000000          0.000000   
17  0.000000             0.000000             0.000000          0.000000   
18  0.000000             0.000000             0.000000          0.000000   
19  0.000000             0.000000             0.000000          0.000000   
20  0.000000             0.000000             0.027027          0.000000   
21  0.000000             0.000000             0.000000          0.000000   
22  0.000000             0.000000             0.000000          0.000000   

    Athletics & Sports    Bakery      Bank       Bar  Baseball Field  \
0                  0.0  0.000000  0.055556  0.000000        0.000000   
1                  0.0  0.000000  0.250000  0.000000        0.000000   
2                  0.0  0.000000  0.000000  0.000000        0.000000   
3                  0.0  0.000000  0.000000  0.000000        0.000000   
4                  0.0  0.000000  0.000000  0.000000        0.200000   
5    

In [45]:
northyork_grouped.shape

(23, 112)

In [46]:
#print each neighborhood along with the top 5 most common venues
num_top_venues = 5

for hood in northyork_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = northyork_grouped[northyork_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor, Downsview North, Wilson Heights----
                 venue  freq
0          Coffee Shop  0.11
1                Diner  0.06
2        Shopping Mall  0.06
3           Restaurant  0.06
4  Fried Chicken Joint  0.06


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Bank  0.25
2                 Café  0.25
3  Japanese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                  venue  freq
0  Fast Food Restaurant  0.08
1    Italian Restaurant  0.08
2           Coffee Shop  0.08
3              Pharmacy  0.04
4        Sandwich Place  0.04


----CFB Toronto, Downsview East----
               venue  freq
0            Airport  0.33
1               Park  0.33
2  Other Repair Shop  0.33
3                Gym  0.00
4          Pet Store  0.00


----Don Mills North----
                  venue  freq
0  Gym / Fitness Center   0.2
1  Caribbean Restaurant   0.2
2                  Café   0.2


In [47]:
#turn top5 for each neighborhood into dataframe
#write a function to sort venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [72]:
#create the new dataframe and display the top 10 venues for each neighborhood
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = northyork_grouped['Neighborhood']

for ind in np.arange(northyork_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(northyork_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Bathurst Manor, Downsview North, Wilson Heights           Coffee Shop   
1                                  Bayview Village    Chinese Restaurant   
2                Bedford Park, Lawrence Manor East           Coffee Shop   
3                      CFB Toronto, Downsview East                  Park   
4                                  Don Mills North  Gym / Fitness Center   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0           Supermarket  Middle Eastern Restaurant    Frozen Yogurt Shop   
1                  Café                       Bank   Japanese Restaurant   
2  Fast Food Restaurant         Italian Restaurant         Grocery Store   
3               Airport          Other Repair Shop               Dog Run   
4  Caribbean Restaurant                       Café        Baseball Field   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                    Pharmacy           Pizza Place         Deli / Bodega   
1               Women's Store     Electronics Store        Cosmetics Shop   
2           Indian Restaurant          Liquor Store          Cupcake Shop   
3  Construction & Landscaping     Convenience Store        Cosmetics Shop   
4         Japanese Restaurant         Women's Store     Electronics Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant           Bridal Shop         Sandwich Place  
1          Cupcake Shop         Deli / Bodega       Department Store  
2                  Café              Pharmacy            Pizza Place  
3          Cupcake Shop         Deli / Bodega       Department Store  
4        Cosmetics Shop          Cupcake Shop          Deli / Bodega

#### Cluster neighborhood

In [73]:
# set number of clusters
kclusters = 6

northyork_grouped_clustering = northyork_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(northyork_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 5, 3, 1, 5, 5, 3, 3, 1, 4], dtype=int32)

In [74]:
#Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

northyork_merged = northyork_data

# merge neighborhoods_venues_sorted with northyork_data to add latitude/longitude for each neighborhood
northyork_merged = northyork_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

northyork_merged.head() # check the last columns!

PostCode     Borough                  Neighborhood   Latitude  Longitude  \
0      M2H  North York             Hillcrest Village  43.803762 -79.363452   
1      M2J  North York  Fairview, Henry Farm, Oriole  43.778517 -79.346556   
2      M2K  North York               Bayview Village  43.786947 -79.385975   
3      M2L  North York      Silver Hills, York Mills  43.757490 -79.374714   
4      M2M  North York       Newtonbrook, Willowdale  43.789053 -79.408493   

   Cluster Labels 1st Most Common Venue     2nd Most Common Venue  \
0             3.0               Dog Run  Mediterranean Restaurant   
1             3.0        Clothing Store      Fast Food Restaurant   
2             5.0    Chinese Restaurant                      Café   
3             2.0             Cafeteria             Women's Store   
4             NaN                   NaN                       NaN   

  3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
0                  Pool                 Golf Course   Fried Chicken Joint   
1           Coffee Shop                  Restaurant            Kids Store   
2                  Bank         Japanese Restaurant         Women's Store   
3           Coffee Shop  Construction & Landscaping     Convenience Store   
4                   NaN                         NaN                   NaN   

  6th Most Common Venue    7th Most Common Venue       8th Most Common Venue  \
0    Dim Sum Restaurant  Comfort Food Restaurant  Construction & Landscaping   
1      Toy / Game Store                 Tea Room               Metro Station   
2     Electronics Store           Cosmetics Shop                Cupcake Shop   
3        Cosmetics Shop             Cupcake Shop               Deli / Bodega   
4                   NaN                      NaN                         NaN   

    9th Most Common Venue 10th Most Common Venue  
0  Furniture / Home Store      Convenience Store  
1                  Bakery    Japanese Restaurant  
2           Deli / Bodega       Department Store  
3        Department Store           Dessert Shop  
4                     NaN                    NaN

In [75]:
#northyork_merged data has one row with NaN values and drop it
northyork_merged.dropna(axis=0, inplace=True)
northyork_merged.shape

(23, 16)

In [76]:
#visualize resulting venues
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(northyork_merged['Latitude'], northyork_merged['Longitude'], northyork_merged['Neighborhood'], northyork_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine Cluster

In [77]:
#Cluster 1
northyork_merged.loc[northyork_merged['Cluster Labels'] == 0, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
22  North York             0.0              Pharmacy   Empanada Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22           Pizza Place    Frozen Yogurt Shop                 Diner   

   6th Most Common Venue    7th Most Common Venue       8th Most Common Venue  \
22           Coffee Shop  Comfort Food Restaurant  Construction & Landscaping   

   9th Most Common Venue 10th Most Common Venue  
22     Convenience Store         Cosmetics Shop

In [78]:
#Cluster 2
northyork_merged.loc[northyork_merged['Cluster Labels'] == 1, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
6   North York             1.0                  Park   
8   North York             1.0                  Park   
13  North York             1.0                  Park   
20  North York             1.0           Pizza Place   
21  North York             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6            Convenience Store                  Bank                   Bar   
8            Food & Drink Shop  Fast Food Restaurant         Women's Store   
13                     Airport     Other Repair Shop               Dog Run   
20                         Pub   Japanese Restaurant                  Park   
21  Construction & Landscaping         Deli / Bodega                Bakery   

         5th Most Common Venue       6th Most Common Venue  \
6            Electronics Store  Construction & Landscaping   
8                      Dog Run  Construction & Landscaping   
13  Construction & Landscaping           Convenience Store   
20               Women's Store                       Diner   
21            Basketball Court           Electronics Store   

      7th Most Common Venue       8th Most Common Venue 9th Most Common Venue  \
6            Cosmetics Shop                Cupcake Shop         Deli / Bodega   
8         Convenience Store              Cosmetics Shop          Cupcake Shop   
13           Cosmetics Shop                Cupcake Shop         Deli / Bodega   
20  Comfort Food Restaurant  Construction & Landscaping     Convenience Store   
21        Convenience Store              Cosmetics Shop          Cupcake Shop   

   10th Most Common Venue  
6        Department Store  
8           Deli / Bodega  
13       Department Store  
20         Cosmetics Shop  
21       Department Store

In [79]:
#Cluster 3
northyork_merged.loc[northyork_merged['Cluster Labels'] == 2, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3  North York             2.0             Cafeteria         Women's Store   

  3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
3           Coffee Shop  Construction & Landscaping     Convenience Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3        Cosmetics Shop          Cupcake Shop         Deli / Bodega   

  9th Most Common Venue 10th Most Common Venue  
3      Department Store           Dessert Shop

In [82]:
#Cluster 4
northyork_merged.loc[northyork_merged['Cluster Labels'] == 3, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
0   North York             3.0               Dog Run   
1   North York             3.0        Clothing Store   
5   North York             3.0      Ramen Restaurant   
7   North York             3.0         Grocery Store   
10  North York             3.0                   Gym   
11  North York             3.0           Coffee Shop   
12  North York             3.0           Coffee Shop   
14  North York             3.0         Grocery Store   
16  North York             3.0         Grocery Store   
17  North York             3.0          Hockey Arena   
18  North York             3.0           Coffee Shop   
19  North York             3.0        Clothing Store   

       2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0   Mediterranean Restaurant                       Pool           Golf Course   
1       Fast Food Restaurant                Coffee Shop            Restaurant   
5                Coffee Shop           Sushi Restaurant            Restaurant   
7                Pizza Place             Discount Store           Coffee Shop   
10          Asian Restaurant                Coffee Shop            Beer Store   
11               Supermarket  Middle Eastern Restaurant    Frozen Yogurt Shop   
12        Miscellaneous Shop       Caribbean Restaurant         Metro Station   
14                      Bank              Shopping Mall     Electronics Store   
16      Gym / Fitness Center         Athletics & Sports        Discount Store   
17     Portuguese Restaurant               Intersection           Coffee Shop   
18      Fast Food Restaurant         Italian Restaurant         Grocery Store   
19    Furniture / Home Store              Women's Store    Miscellaneous Shop   

         5th Most Common Venue 6th Most Common Venue  \
0          Fried Chicken Joint    Dim Sum Restaurant   
1                   Kids Store      Toy / Game Store   
5               Sandwich Place   Japanese Restaurant   
7                     Pharmacy      Asian Restaurant   
10               Grocery Store            Smoke Shop   
11                    Pharmacy           Pizza Place   
12              Massage Studio                   Bar   
14  Construction & Landscaping     Convenience Store   
16                Liquor Store     Electronics Store   
17               Women's Store               Dog Run   
18           Indian Restaurant          Liquor Store   
19                    Boutique           Coffee Shop   

         7th Most Common Venue       8th Most Common Venue  \
0      Comfort Food Restaurant  Construction & Landscaping   
1                     Tea Room               Metro Station   
5                         Café                   Pet Store   
7            Electronics Store           Convenience Store   
10        Fast Food Restaurant          Italian Restaurant   
11               Deli / Bodega                  Restaurant   
12          Falafel Restaurant              Discount Store   
14              Cosmetics Shop                Cupcake Shop   
16           Convenience Store              Cosmetics Shop   
17  Construction & Landscaping           Convenience Store   
18                Cupcake Shop                        Café   
19                 Event Space            Fraternity House   

     9th Most Common Venue 10th Most Common Venue  
0   Furniture / Home Store      Convenience Store  
1                   Bakery    Japanese Restaurant  
5          Bubble Tea Shop                  Plaza  
7           Cosmetics Shop           Cupcake Shop  
10     Japanese Restaurant         Discount Store  
11             Bridal Shop         Sandwich Place  
12          Cosmetics Shop           Cupcake Shop  
14           Deli / Bodega       Department Store  
16            Cupcake Shop          Deli / Bodega  
17          Cosmetics Shop           Cupcake Shop  
18                Pharmacy            Pizza Place  
19       Accessories Store  Vietnamese Restaurant

In [83]:
#Cluster 5
northyork_merged.loc[northyork_merged['Cluster Labels'] == 4, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
23  North York             4.0        Baseball Field         Women's Store   

   3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
23     Electronics Store  Construction & Landscaping     Convenience Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
23        Cosmetics Shop          Cupcake Shop         Deli / Bodega   

   9th Most Common Venue 10th Most Common Venue  
23      Department Store           Dessert Shop

In [84]:
#Cluster 6
northyork_merged.loc[northyork_merged['Cluster Labels'] == 5, northyork_merged.columns[[1] + list(range(5, northyork_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2   North York             5.0    Chinese Restaurant                  Café   
9   North York             5.0  Gym / Fitness Center  Caribbean Restaurant   
15  North York             5.0     Korean Restaurant          Home Service   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                   Bank   Japanese Restaurant         Women's Store   
9                   Café        Baseball Field   Japanese Restaurant   
15            Food Truck        Baseball Field         Women's Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2      Electronics Store        Cosmetics Shop          Cupcake Shop   
9          Women's Store     Electronics Store        Cosmetics Shop   
15     Convenience Store        Cosmetics Shop          Cupcake Shop   

   9th Most Common Venue 10th Most Common Venue  
2          Deli / Bodega       Department Store  
9           Cupcake Shop          Deli / Bodega  
15         Deli / Bodega       Department Store